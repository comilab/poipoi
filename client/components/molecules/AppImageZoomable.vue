<template lang="pug">
.inline-block
  .zoomable-area.relative(
    @click="zoomed = true",
    data-testid="zoomable-area"
  )
    img.select-none.pointer-events-none(
      :src="src",
      :class="imgClass",
      data-testid="zoomable-area-img"
    )
    .caution.absolute.right-0.bottom-0.text-white.text-xs.p-2 無断転載禁止

  transition(
    enter-class="opacity-0",
    leave-to-class="opacity-0"
  )
    .zoomed-area.flex.fixed.inset-0.overflow-auto.bg-white.bg-opacity-75.z-50.transition-opacity.duration-300(
      v-if="zoomed",
      @click="zoomed = false",
      data-testid="zoomed-area"
    )
      .m-auto
        img.max-w-none.select-none.pointer-events-none.p-4(
          :src="largeSrc"
        )
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs } from 'nuxt-composition-api'

import { prop as classProp } from '~/composables/use-class-prop'

export default defineComponent({
  props: {
    src: {
      type: String,
      required: true
    },
    largeSrc: {
      type: String,
      required: true
    },
    imgClass: {
      ...classProp
    }
  },
  setup () {
    const localState = reactive({
      zoomed: false
    })

    return {
      ...toRefs(localState)
    }
  }
})
</script>

<style lang="postcss" scoped>
.zoomable-area {
  cursor: zoom-in;
}

.zoomed-area {
  cursor: zoom-out;
}

.caution {
  text-shadow: 0 0 5px #000;
}
</style>
