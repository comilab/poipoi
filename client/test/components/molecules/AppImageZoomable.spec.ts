import { render, RenderOptions, fireEvent } from '@testing-library/vue'

import AppImageZoomable from '~/components/molecules/AppImageZoomable.vue'

describe('components/molecules/AppImageZoomable', () => {
  let options: RenderOptions<AppImageZoomable>
  let props: any

  beforeEach(() => {
    props = {
      src: 'small.jpeg',
      largeSrc: 'large.jpeg'
    }

    options = {
      props
    }
  })

  test('renders correctly', () => {
    const { container, queryByTestId } = render(AppImageZoomable, options)

    expect(container).toMatchSnapshot()
    expect(queryByTestId('zoomable-area')).toBeVisible()
    expect(queryByTestId('zoomed-area')).not.toBeInTheDocument()
  })

  describe('props', () => {
    describe('imgClass', () => {
      test('img要素にクラスが付与される', () => {
        props.imgClass = 'foo'
        const { container, queryByTestId } = render(AppImageZoomable, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('zoomable-area-img')).toHaveClass('foo')
      })
    })
  })

  test('zoomable-areaをクリックするとズームする', async () => {
    const { container, getByTestId, queryByTestId } = render(AppImageZoomable, options)
    await fireEvent.click(getByTestId('zoomable-area'))

    expect(container).toMatchSnapshot()
    expect(queryByTestId('zoomed-area')).toBeVisible()
  })

  test('zoomed-areaをクリックするとズームアウトする', async () => {
    const { container, getByTestId, queryByTestId } = render(AppImageZoomable, options)
    await fireEvent.click(getByTestId('zoomable-area'))
    await fireEvent.click(getByTestId('zoomed-area'))

    expect(container).toMatchSnapshot()
    expect(queryByTestId('zoomed-area')).not.toBeInTheDocument()
  })
})
