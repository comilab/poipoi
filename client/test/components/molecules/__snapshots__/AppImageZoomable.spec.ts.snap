// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppImageZoomable props imgClass img要素にクラスが付与される 1`] = `
<div>
  <div
    class="inline-block"
  >
    <div
      class="zoomable-area relative"
      data-testid="zoomable-area"
    >
      <img
        class="select-none pointer-events-none foo"
        data-testid="zoomable-area-img"
        src="small.jpeg"
      />
      <div
        class="caution absolute right-0 bottom-0 text-white text-xs p-2"
      >
        無断転載禁止
      </div>
    </div>
    <transition-stub
      enterclass="opacity-0"
      leavetoclass="opacity-0"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`components/molecules/AppImageZoomable renders correctly 1`] = `
<div>
  <div
    class="inline-block"
  >
    <div
      class="zoomable-area relative"
      data-testid="zoomable-area"
    >
      <img
        class="select-none pointer-events-none"
        data-testid="zoomable-area-img"
        src="small.jpeg"
      />
      <div
        class="caution absolute right-0 bottom-0 text-white text-xs p-2"
      >
        無断転載禁止
      </div>
    </div>
    <transition-stub
      enterclass="opacity-0"
      leavetoclass="opacity-0"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`components/molecules/AppImageZoomable zoomable-areaをクリックするとズームする 1`] = `
<div>
  <div
    class="inline-block"
  >
    <div
      class="zoomable-area relative"
      data-testid="zoomable-area"
    >
      <img
        class="select-none pointer-events-none"
        data-testid="zoomable-area-img"
        src="small.jpeg"
      />
      <div
        class="caution absolute right-0 bottom-0 text-white text-xs p-2"
      >
        無断転載禁止
      </div>
    </div>
    <transition-stub
      enterclass="opacity-0"
      leavetoclass="opacity-0"
    >
      <div
        class="zoomed-area flex fixed inset-0 overflow-auto bg-white bg-opacity-75 z-50 transition-opacity duration-300"
        data-testid="zoomed-area"
      >
        <div
          class="m-auto"
        >
          <img
            class="max-w-none select-none pointer-events-none p-4"
            src="large.jpeg"
          />
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`components/molecules/AppImageZoomable zoomed-areaをクリックするとズームアウトする 1`] = `
<div>
  <div
    class="inline-block"
  >
    <div
      class="zoomable-area relative"
      data-testid="zoomable-area"
    >
      <img
        class="select-none pointer-events-none"
        data-testid="zoomable-area-img"
        src="small.jpeg"
      />
      <div
        class="caution absolute right-0 bottom-0 text-white text-xs p-2"
      >
        無断転載禁止
      </div>
    </div>
    <transition-stub
      enterclass="opacity-0"
      leavetoclass="opacity-0"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;
